xdsl/pattern_rewriter.py

```diff
@@ -439,6 +457,140 @@ def impl(self: _RewritePatternT, op: Operation, rewriter: PatternRewriter) -> No
     return impl
 
 
+@dataclass
+class TypeConversionPattern(RewritePattern):
+    """
+    Base pattern for type conversion. It is supposed to be inherited from, then one can
+    implement `convert_type` to define the conversion.
+
+    It will convert an Operations' result types, dictionary attributes, and block arguments.
+
+    One can use `@attr_type_rewrite_pattern` on this defined method to automatically filter
+    on the Attribute type used.
+
+    This base pattern defines two flags:
+
+    - `recursive` (defaulting to True): recurse over structured attributes to convert
+      parameters.
+      e.g. a recusrive `i32` to `index` conversion will convert `vector<i32>` to
+      `vector<index>`.
+
+    - `ops` (defaulting to any Operation) is a tuple of Operation types on which to apply
+      the defined attribute conversion.
+    """
+
+    recursive: bool = True
+    ops: tuple[type[Operation]] | None = None
+
+    @abstractmethod
+    def convert_type(self, typ: Attribute, /) -> Attribute | None:
+        """
+        The method to implement to define a TypeConversionPattern
+
+        This defines how the input Attribute should be converted.
+        It allows returning None, meaning "this attribute should not
+        be converted".
+        """
+        ...
+
+    @final
+    def _convert_type_rec(self, typ: Attribute) -> Attribute | None:
+        """
+        Provided recursion over structed/parameterized Attributes.
+        """
+        inp = typ
+        if self.recursive:
+            if isinstance(typ, ParametrizedAttribute):
+                parameters = list(
+                    self._convert_type_rec(p) or p for p in typ.parameters
+                )
+                inp = type(typ).new(parameters)
+            if isa(typ, ArrayAttr[Attribute]):
+                parameters = tuple(self._convert_type_rec(p) or p for p in typ)
+                inp = type(typ).new(parameters)
+        return self.convert_type(inp) or inp
```

Comment:
Should check explicitly for None, as you can return an instance of a type here that has a custom `__bool__` method returning false (e.g. when it's empty), which would trigger this thing. Instead you should just use `is None`.